src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonAdvancedConverter.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json.Serialization.Metadata;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal abstract class JsonAdvancedConverter<T> : JsonResumableConverter<T>
+    {
+        internal virtual bool IsJsonDictionaryConverter => false;
+
+        internal virtual bool OnWriteResume(Utf8JsonWriter writer, T dictionary, JsonSerializerOptions options, ref WriteStack state)
+        {
+            Debug.Fail("We should have overriden it if we got here");
+            return true;
+        }
+    }
+
+    /// <summary>
+    /// Base class for object, enumerable and dictionary converters
+    /// </summary>
+    internal abstract class JsonAdvancedConverter<T, IntermediateType> : JsonAdvancedConverter<T>
+    {
+        private protected virtual bool TryCreateObject(ref Utf8JsonReader reader, JsonTypeInfo jsonTypeInfo, scoped ref ReadStack state, [NotNullWhen(true)] out IntermediateType? obj)
+        {
+            // TODO: generalize exceptions
+            if (jsonTypeInfo.CreateObject == null)
+            {
+                // The contract model was not able to produce a default constructor for two possible reasons:
+                // 1. Either the declared collection type is abstract and cannot be instantiated.
+                // 2. The collection type does not specify a default constructor.
+                if (TypeToConvert.IsAbstract || TypeToConvert.IsInterface)
+                {
+                    ThrowHelper.ThrowNotSupportedException_CannotPopulateCollection(TypeToConvert, ref reader, ref state);
+                }
+                else
+                {
+                    ThrowHelper.ThrowNotSupportedException_DeserializeNoConstructor(TypeToConvert, ref reader, ref state);
+                }
+            }
+
+            obj = (IntermediateType)jsonTypeInfo.CreateObject();
+
+            if (IsReadOnly(obj))
+            {
+                // TODO: not possible for objects, might be a good idea to still generalize exception
```

Comment:
This is problematic. Object deserialization is more common than collection deserialization, wouldn't want to spend CPU cycles checking for something that is always true.

Reply:
that's really a single check per object, I was thinking that this type could eventually be public and making this generalized (if we do delegates that will also be true)